contrib/implements/internal/implements/gosrc.go

```diff
@@ -9,44 +10,126 @@ import (
 	"path"
 	"regexp"
 	"strings"
+	"unicode"
 )
 
+// ignoreCCalls is a map of calls from "C" that we know we can just ignore.
+// ignoring this stuff makes the number of tracked entries smaller, fewer logs,
+// and less stuff to iterate over in other code.
+var ignoreCCalls = map[string]bool{
+	// some special cgo calls
+	"CString":  true,
+	"CBytes":   true,
+	"GoString": true,
+	"GoBytes":  true,
+	// common utility functions
+	"free": true,
+	// common types
+	"int":       true,
+	"int64_t":   true,
+	"uint64_t":  true,
+	"size_t":    true,
+	"ssize_t":   true,
+	"uintptr_t": true,
+}
+
+type goFunction struct {
+	shortName       string
+	fullName        string
+	comment         string
+	implementsCFunc string
+	callsCFunc      string
+	isDeprecated    bool
+	isPreview       bool
+
+	endPos token.Pos
+}
+
 type visitor struct {
-	inFunction *ast.FuncDecl
+	currentFunc *goFunction
 
-	callMap map[string]string
-	docMap  map[string]string
+	callMap    map[string]*goFunction
+	docMap     map[string]*goFunction
+	deprecated []*goFunction
+	preview    []*goFunction
+	stable     []*goFunction
 }
 
 func newVisitor() *visitor {
 	return &visitor{
-		callMap: map[string]string{},
-		docMap:  map[string]string{},
+		callMap:    map[string]*goFunction{},
+		docMap:     map[string]*goFunction{},
+		deprecated: []*goFunction{},
+		preview:    []*goFunction{},
+		stable:     []*goFunction{},
+	}
+}
+
+func funcDeclFullName(fdec *ast.FuncDecl) string {
+	if fdec.Recv == nil {
+		return fdec.Name.Name
+	}
+	if len(fdec.Recv.List) != 1 {
+		return fdec.Name.Name
 	}
+	typeName := "UNKNOWN!"
+	switch t := fdec.Recv.List[0].Type.(type) {
+	case *ast.StarExpr:
+		typeName = t.X.(*ast.Ident).Name
+	case *ast.Ident:
+		typeName = t.Name
+	}
+	return fmt.Sprintf("%s.%s", typeName, fdec.Name.Name)
 }
 
-func (v *visitor) checkDocImplements(fdec *ast.FuncDecl) {
-	dtext := fdec.Doc.Text()
-	lines := strings.Split(dtext, "\n")
+func readDocComment(fdec *ast.FuncDecl, gfunc *goFunction) {
+	gfunc.comment = fdec.Doc.Text()
+	lines := strings.Split(gfunc.comment, "\n")
 	for i := range lines {
+		if strings.Contains(lines[i], "DEPRECATED") {
+			gfunc.isDeprecated = true
+			logger.Printf("marked deprecated: %s\n", fdec.Name.Name)
+		}
+		if strings.Contains(lines[i], "PREVIEW") {
+			gfunc.isPreview = true
+			logger.Printf("marked preview: %s\n", fdec.Name.Name)
+		}
+
 		if lines[i] == "Implements:" {
 			cfunc := cfuncFromComment(lines[i+1])
 			if cfunc == "" {
 				return
 			}
-			v.docMap[cfunc] = fdec.Name.Name
-			logger.Printf("updated %s in doc map\n", cfunc)
+			gfunc.implementsCFunc = cfunc
+			logger.Printf("implements c function %s: %s\n", fdec.Name.Name, cfunc)
 		}
 	}
 }
 
+func isPublic(gfunc *goFunction) bool {
+	if len(gfunc.shortName) == 0 {
+		return false
+	}
+	// is there a better way to the first rune?
+	var firstRune rune
+	for _, r := range gfunc.shortName {
+		firstRune = r
+		break
+	}
```

Comment:
According to https://stackoverflow.com/a/30263910 you can also use the unicode/utf8 package:

```go
firstRune, _ := utf8.DecodeRuneInString(gfunc.shortName)
```

Reply:
Actually, you don't need this function: https://pkg.go.dev/go/ast#IsExported

Reply:
I swear I looked for that exact function and didn't find it!

Reply:
In German we say: "four eyes can see better than two". 👀👀